asr-python-adapter-finetune-am-conformer-nemo.ipynb

```diff
@@ -0,0 +1,186 @@
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# What are Adapters?\n",
+    "\n",
+    "Adapters are trainable neural network modules that are attached to pretrained models, such that we freeze the weights of the original model and only train the adapter parameters. This reduces the amount of data required to customize a model substantially, while imposing the limitation that the model's vocabulary cannot be changed.\n",
+    "\n",
+    "In short,\n",
+    "\n",
+    "- Adapter modules form a residual bridge over the output of each layer they adapt, such that the model's original performance is not lost. \n",
+    "- The original parameters of the model are frozen in their entirety - so that we can recover the original model by disabling all adapters.\n",
+    "- We train only the new adapter parameters (an insignificant fraction of the total number of parameters). This allows fast experimentation with very little data and compute.\n",
+    "\n",
+    "-----\n",
+    "\n",
+    "Adapters are a straightforward concept - as shown by the diagram below. At their simplest, they are residual Feedforward layers that compress the input dimension ($D$) to a small bottleneck dimension ($H$), such that $R^D \\text{->} R^H$, compute an activation (such as ReLU), finally mapping $R^H \\text{->} R^D$ with another Feedforward layer. This output is then added to the input via a residual connection.\n",
+    "\n",
+    "<div align=\"center\">\n",
+    "  <img src=\"https://mermaid.ink/img/pako:eNptkLFqwzAQhl9F3ORAPDSjA4EUx6RgXEjbycpwWOdG1JaMfEoakrx7ZcfpUKrlxH_fz4d0gcoqggTqxp6qAzoW76k0Ipx1-WI6z3sRxyuRF1GOZ3KisK6d3YG8GFdZ9hRJeLbMDRmqvkRGpDLrTuiUiEWUigBtlyIVqzBnEqZ66I39dcX6iKytKXeUf-wn-286QoFeBMvmu0PTD-EfyXaQpP9JFmP_1XN4S3kfD8W4ue6o18pjc52gYQlzaMm1qFX4msuQSOADtSQhCdfaOupZgjS3QPpOIdNGabYOkhqbnuaAnu3b2VSQsPP0gFKNnw7bibr9AJkZdXU\" height=100% />\n",
+    "</div>\n",
+    "\n",
+    "-----"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Advantages and Limitations of Adapter Training\n",
+    "\n",
+    "Adapters can be used with limited amounts of training data and compute budget, but they impose the restriction that the model's original vocabulary must be used - therefore a new characterset vocabulary / tokenizer cannot be used.\n",
+    "\n",
+    "Please refer to the **Advantages of Adapters** and **Limitations of Adapters** sections in the [NeMo Tutorial](#asr-with-adapters) for further details."
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Preparing the Acoustic Encoder for Adapter Training\n",
+    "\n",
+    "Pre-trained models do not automatically support Adapter modules, so we must prepare the acoustic encoder prior to creating the model.\n",
+    "The steps for this procedure are detailed in the **Prepare the \"base\" model** section of the [NeMo Tutorial](#asr-with-adapters)."
```

Comment:
```suggestion
    "The steps for this procedure are detailed in the **Prepare the \"base\" model** section of the [NeMo tutorial](#asr-with-adapters)."
```

Reply:
I'm assuming in the previous comment and this, we cannot directly link to a particular section within a tutorial? Or is there any fear that these sections may be removed and break our links...